### ISSUE TITLE
Kubectl - Azure AD authentication adds excessive 'spn:' to SPN

### ISSUE THREAD

**Issue by alastairpat** [kubernetes/kubernetes#89672]
**What happened**:
When using kubectl, authentication fails with the below error
`E0331 10:18:17.205561   46266 azure.go:154] Failed to acquire a token: refreshing the expired token: refreshing token: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_resource","error_description":"AADSTS500011: The resource principal named spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:**spn** was not found in the tenant named **tenant ID**. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.\r\nTrace ID: **trace ID**\r\nCorrelation ID: **correlation ID**\r\nTimestamp: 2020-03-30 23:18:17Z","error_codes":[500011],"timestamp":"2020-03-30 23:18:17Z","trace_id":"**trace ID**","correlation_id":"**correlation ID**","error_uri":"https://login.microsoftonline.com/error?code=500011"}`

For some reason kubectl is adding a lot of `spn:`s to the start of the SPN - this can be observed in ~/.kube/config:
```users:
- name: azure-active-directory
  user:
    auth-provider:
      config:
        apiserver-id: spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:spn:**SPN**
```

**What you expected to happen**:
Authentication works as in previous versions of Kubectl

**How to reproduce it (as minimally and precisely as possible)**:
Create a cluster using Azure AD authentication. Authenticate from a client and wait until the token expires. Use Kubectl and have it automatically refresh the token; it will fail due to the extra 'spn:' entries.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration: on-premise
- OS (e.g: `cat /etc/os-release`): macOS 10.14.6
- Kernel (e.g. `uname -a`): Darwin Hostname 18.7.0 Darwin Kernel Version 18.7.0: Mon Feb 10 21:08:45 PST 2020; root:xnu-4903.278.28~1/RELEASE_X86_64 x86_64
- Install tools: N/A
- Network plugin and version (if this is a network-related bug): N/A
- Others:

**Comment by alastairpat** [comment]
Kubectl version 1.18.0
Kubernetes version 1.16.6

**Comment by weinong** [comment]
@alastairpat can you confirm whether your original kubeconfig already has `spn:` in apiserver-id or not? 
in the mean time, you may try https://github.com/Azure/kubelogin
it can convert your kubeconfig from azure module to exec module and do the token thing outside kubectl

**Comment by UnwashedMeme** [comment]
```
$ sed -i -E 's/(spn:)+//' ~/.kube/config
$ grep apiserver ~/.kube/config          
        apiserver-id: fb55499d-d745-469d-9f3d-dbb31734e32c
$ sed -ie /token/d ~/.kube/config
$ k cluster-info > /dev/null             
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code ... to authenticate.
$ grep apiserver ~/.kube/config  
        apiserver-id: fb55499d-d745-469d-9f3d-dbb31734e32c
$ k cluster-info > /dev/null   
$ grep apiserver ~/.kube/config
        apiserver-id: spn:fb55499d-d745-469d-9f3d-dbb31734e32c
$ k cluster-info > /dev/null   
$ grep apiserver ~/.kube/config
        apiserver-id: spn:spn:fb55499d-d745-469d-9f3d-dbb31734e32c
$ k cluster-info > /dev/null   
$ grep apiserver ~/.kube/config
        apiserver-id: spn:spn:spn:fb55499d-d745-469d-9f3d-dbb31734e32c
```
At this point if I remove the tokens and try to reauth i get the error described above.

I've tried it both ways with apiserver-id having an `spn:` initially and not, both lead to the same behavior in the end.

**Comment by weinong** [comment]
@UnwashedMeme i know what's going on now. will prepare a fix.

**Comment by kwaazaar** [comment]
I'm seeing this with kubectl 1.17.3 and kubernetes 1.16.7. It's unclear to me what versions are impacted and in what version it is fixed, since only version 1.18 is mentioned.

**Comment by antoinedeschenes** [comment]
This still happens on kubectl 1.16.12 and 1.17.8

**Comment by AlekseyBekirov** [comment]
This still happens on kubectl v1.18.5

**Comment by bitsofinfo** [comment]
same

**Comment by weinong** [comment]
can you guys share your kubectl version --client and kubeconfig with me? you can find me on kubernetes slack. the slack user name is same as my github handle.

**Comment by alexbekir** [comment]
Version 1.17.6 works fine.
```
kubectl version --client=true
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.6", GitCommit:"d32e40e20d167e103faf894261614c5b45c44198", GitTreeState:"clean", BuildDate:"2020-05-20T13:16:24Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"darwin/amd64"}
```

**Comment by whitlaaa** [comment]
I'm also still seeing this on v1.18.6

```
$ kubectl version --client
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:58:53Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by whitlaaa** [comment]
Since I was mistaken thinking `kubectl` was still causing this for me, I figured I would just drop a note for others in case it's helpful. Turns out my problem was actually due to `helm` and looks to have been resolved as of the v3.3.0 release.

https://github.com/helm/helm/issues/8104

**Comment by audunsolemdal** [comment]
This issue has now reappeared for me

AKS 1.19.0

Authentication works, but it spams frequently.

```
❯ helm version
version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"clean", GoVersion:"go1.14.9"}
⚡ audun@PF1N15A8  C:\appl\repos                                                                                                                  [13:27]
❯ ^C
⚡ audun@PF1N15A8  C:\appl\repos                                                                                                                  [13:27]
❯ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T18:49:11Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```
